# Phase 3 Project

Our data science team has been contracted by a fairly new up and coming motorcycle manufacturer that is struggling to find their new formula for a sales success. Being a company of motorcycle riders, they have tasked our team to come up with a plan to replicate the successes of platforms like the Yamaha MT-07 and the Suzuki SV650. We will use the 'all_bikes_curated' dataset from Kaggle, curated by Emmanuel F. Werr


<img src="pictures/092220-2021-bmw-m1000rr-f.webp"  />
<center>2022 BMW M1000RR</center>

In [1]:
# Import warnings 
import warnings
warnings.filterwarnings('ignore')

# Import pandas to read our data
import pandas as pd
# 'all_bikez_curated.csv', with a 'z'
df = pd.read_csv('all_bikez_curated.csv')
# Show our data 
df.head()

Brand                Model  Year                   Category  Rating  \
0  acabion      da vinci 650-vi  2011  Prototype / concept model     3.2   
1  acabion              gtbo 55  2007                      Sport     2.6   
2  acabion  gtbo 600 daytona-vi  2011  Prototype / concept model     3.5   
3  acabion  gtbo 600 daytona-vi  2021  Prototype / concept model     NaN   
4  acabion              gtbo 70  2007  Prototype / concept model     3.1   

   Displacement (ccm)  Power (hp)  Torque (Nm) Engine cylinder Engine stroke  \
0                 NaN       804.0          NaN        Electric      Electric   
1              1300.0       541.0        420.0    In-line four   four-stroke   
2                 NaN       536.0          NaN        Electric      Electric   
3                 NaN       536.0          NaN        Electric      Electric   
4              1300.0       689.0        490.0    In-line four   four-stroke   

   ... Dry weight (kg)  Wheelbase (mm) Seat height (mm)  Front brakes  \
0  ...           420.0             NaN              NaN   Single disc   
1  ...           360.0             NaN              NaN           NaN   
2  ...           420.0             NaN              NaN   Single disc   
3  ...           420.0             NaN              NaN   Single disc   
4  ...           300.0             NaN              NaN           NaN   

   Rear brakes Front tire Rear tire Front suspension  Rear suspension  \
0  Single disc        NaN       NaN              NaN              NaN   
1          NaN        NaN       NaN              NaN              NaN   
2  Single disc        NaN       NaN              NaN              NaN   
3  Single disc        NaN       NaN              NaN              NaN   
4          NaN        NaN       NaN              NaN              NaN   

   Color options  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4   Custom made.  

[5 rows x 28 columns]

# What makes a rating, good? 

In [2]:
# .unique() will give you all of the potential scores found in the 'Rating' column 
df['Rating'].unique()

array([3.2, 2.6, 3.5, nan, 3.1, 3.9, 2.8, 2.9, 1.9, 3.4, 2.2, 3.3, 3. ,
       2.7, 2.5, 3.7, 2. , 3.6, 3.8, 2.1, 2.4, 4. , 4.1, 2.3, 1.8, 4.2,
       1.7, 4.4, 1.4, 4.3, 4.6, 1.6, 4.5, 1.5])

In [11]:
# Drop nan values from our 'Rating' column
df = df.dropna(subset = ['Rating'])

<div class="alert alert-block alert-info">
Since it would appear that our ratings work on a 5 point system, let's take a look at the motorcycle whose sales performance we would like to replicate: the Yamaha MT-07 and the Suzuki SV650.
</div>

# Yamaha MT-07 & Suzuki SV650

If like me, you are an avid motorcycle rider, you know, have known, or are someone on one of these bikes. 

<center>
<table><tr>
<td> <img src="pictures/yamaha mt07.jfif" alt="Drawing" style="width: 250px;"/> </td>
<td> <img src="pictures/suzuki sv650.jfif" alt="Drawing" style="width: 250px;"/> </td>
</tr></table>
MT-07 and SV650, respectively</center>
<!-- <img src="pictures/yamaha mt07.jfif" />
<img src="pictures/suzuki sv650.jfif" /> -->

In [12]:
# Find all included models that are an MT-07
df_yamaha = df.loc[df['Model'] == 'mt-07']
df_yamaha.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
35463  yamaha  mt-07  2014  Naked bike     3.5               689.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
35463        47.3         87.5            Twin   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  \
35463           175.0          1440.0            815.0   

                   Front brakes               Rear brakes    Front tire  \
35463  Double disc. Hydraulic.   Single disc. Hydraulic.   120/70-ZR17    

          Rear tire  Front suspension                   Rear suspension  \
35463  180/55-ZR17   Telescopic forks  Swingarm, (Link type suspension)   

       Color options  
35463   Black, white  

[1 rows x 28 columns]

In [13]:
# Find all models that have the name SV650
df_suzuki = df.loc[df['Model'] == 'sv650']
df_suzuki.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
30500  suzuki  sv650  2008  Naked bike     3.9               645.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
30500         NaN          NaN              V2   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  Front brakes  \
30500           168.0          1440.0            800.0   Double disc   

       Rear brakes    Front tire     Rear tire  \
30500  Single disc  120/60-ZR17   160/60-ZR17    

                                        Front suspension  \
30500  Telescopic, coil spring, oil damped, fully adj...   

                                  Rear suspension  Color options  
30500  Link-type, 7-way adjustable spring preload     Blue, Gray  

[1 rows x 28 columns]

In [14]:
# Bring in numpy
import numpy as np

In [15]:
# Average of ratings scores for the MT-07
yamaha_mean = np.mean(df_yamaha['Rating'].unique())
print('The average score of all year-model MT-07 in dataset:{}'.format(yamaha_mean))

The average score of all year-model MT-07 in dataset:3.35


In [16]:
# Average of ratings scores for the SV650
suzuki_mean = np.mean(df_suzuki['Rating'].unique())
print('The average score of all year-model SV650 in dataset:{}'.format(suzuki_mean))

The average score of all year-model SV650 in dataset:3.457142857142857


In [17]:
# Nothing fancy here, find the average of the two scores
rating_success = (yamaha_mean + suzuki_mean) / 2
print('Ultimately, this will be our ratings goal:{}'.format(rating_success))

Ultimately, this will be our ratings goal:3.4035714285714285


# Correlations to 'Rating'

Let's take a look at how correlated our categories are to 'Rating'

In [18]:
# How many null values are left in our dataframe?
df.isnull().sum().sum()

114006

In [19]:
# Preview data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21788 entries, 0 to 38461
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                21788 non-null  object 
 1   Model                21784 non-null  object 
 2   Year                 21788 non-null  int64  
 3   Category             21788 non-null  object 
 4   Rating               21788 non-null  float64
 5   Displacement (ccm)   21684 non-null  float64
 6   Power (hp)           14892 non-null  float64
 7   Torque (Nm)          10064 non-null  float64
 8   Engine cylinder      21778 non-null  object 
 9   Engine stroke        21778 non-null  object 
 10  Gearbox              19569 non-null  object 
 11  Bore (mm)            17184 non-null  float64
 12  Stroke (mm)          17184 non-null  object 
 13  Fuel capacity (lts)  18986 non-null  float64
 14  Fuel system          14546 non-null  object 
 15  Fuel control         13444 non-null 

In [20]:
# Remove all remaining null values
df = df.dropna()

In [21]:
# What categories are highly correlated to 'Rating'? 
preds = []
for i in df.corr()['Rating'].index:
    if abs(df.corr()['Rating'][i]) > 0:
        preds.append(i)

In [22]:
# Correlation to 'Rating', all null values removed
df[preds].corr()

Year    Rating  Displacement (ccm)  Power (hp)  \
Year                 1.000000 -0.124482            0.073572    0.047298   
Rating              -0.124482  1.000000            0.264354    0.214983   
Displacement (ccm)   0.073572  0.264354            1.000000    0.661359   
Power (hp)           0.047298  0.214983            0.661359    1.000000   
Torque (Nm)          0.090871  0.265243            0.962469    0.797149   
Bore (mm)            0.073073  0.217992            0.773344    0.572155   
Fuel capacity (lts)  0.030029  0.279262            0.654265    0.566838   
Dry weight (kg)      0.073778  0.209896            0.811809    0.327473   
Wheelbase (mm)       0.072141  0.182865            0.711076    0.324695   
Seat height (mm)    -0.035917  0.038402           -0.255896    0.246490   

                     Torque (Nm)  Bore (mm)  Fuel capacity (lts)  \
Year                    0.090871   0.073073             0.030029   
Rating                  0.265243   0.217992             0.279262   
Displacement (ccm)      0.962469   0.773344             0.654265   
Power (hp)              0.797149   0.572155             0.566838   
Torque (Nm)             1.000000   0.761586             0.675614   
Bore (mm)               0.761586   1.000000             0.502672   
Fuel capacity (lts)     0.675614   0.502672             1.000000   
Dry weight (kg)         0.728967   0.543578             0.667360   
Wheelbase (mm)          0.648251   0.574575             0.597111   
Seat height (mm)       -0.110939  -0.003251             0.033131   

                     Dry weight (kg)  Wheelbase (mm)  Seat height (mm)  
Year                        0.073778        0.072141         -0.035917  
Rating                      0.209896        0.182865          0.038402  
Displacement (ccm)          0.811809        0.711076         -0.255896  
Power (hp)                  0.327473        0.324695          0.246490  
Torque (Nm)                 0.728967        0.648251         -0.110939  
Bore (mm)                   0.543578        0.574575         -0.003251  
Fuel capacity (lts)         0.667360        0.597111          0.033131  
Dry weight (kg)             1.000000        0.805321         -0.442269  
Wheelbase (mm)              0.805321        1.000000         -0.345611  
Seat height (mm)           -0.442269       -0.345611          1.000000

<div class="alert alert-block alert-success">
It appears that there are many 'weak learners' and may be appropriate to use an ensemble method that is good for weak learners, like AdaBoost once we have created our model. 
</div>

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1747 entries, 195 to 38298
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                1747 non-null   object 
 1   Model                1747 non-null   object 
 2   Year                 1747 non-null   int64  
 3   Category             1747 non-null   object 
 4   Rating               1747 non-null   float64
 5   Displacement (ccm)   1747 non-null   float64
 6   Power (hp)           1747 non-null   float64
 7   Torque (Nm)          1747 non-null   float64
 8   Engine cylinder      1747 non-null   object 
 9   Engine stroke        1747 non-null   object 
 10  Gearbox              1747 non-null   object 
 11  Bore (mm)            1747 non-null   float64
 12  Stroke (mm)          1747 non-null   object 
 13  Fuel capacity (lts)  1747 non-null   float64
 14  Fuel system          1747 non-null   object 
 15  Fuel control         1747 non-null 

<div class="alert alert-block alert-info">
All null values have been pruned from our data, let's set our target variable and make a train_test_split. 
</div>

# Binary values for 'Rating'
Now that we have an average score for 'Rating', let's make a new column that will give us binary results, 1 being a rating above 3.4, 0 being a rating below. 

In [24]:
# Let's look for all of the unique values for the column 'Rating'
df['Rating'].unique()

array([2.6, 3.5, 3.6, 3.3, 3.7, 3.4, 2.8, 3.8, 3.1, 3.2, 3.9, 3. , 2.9,
       4. , 2.7, 4.1, 2.5, 4.3, 2.2, 2. , 2.4, 1.8, 4.2, 2.3])

In [25]:
# As detailed above, anything lower than 3.4 is outside of our target
rating_num = np.where(df["Rating"]>=(3.4), 1, 0)

In [26]:
# Our "Rating" column will now return a '1' for bikes that got a rating of 3.4 or above, 0 for scores
# below 3.4
rating_num

array([0, 1, 1, ..., 0, 0, 0])

In [27]:
# In order to attach this data to our current table we will need to make it into a dataframe
data = pd.DataFrame(rating_num, columns = ['Rating_'])
data.head()

Rating_
0        0
1        1
2        1
3        0
4        1

In [28]:
# Preview binary rating counts
data.value_counts()

Rating_
1          1211
0           536
dtype: int64

In [29]:
# Drop 'Rating' column 
df = df.drop(columns = 'Rating', axis =1)
df.head()

Brand              Model  Year          Category  Displacement (ccm)  \
195  aeon           cobra 50  2012               ATV                49.3   
203  aeon   crossland x4 400  2012               ATV               346.0   
226  aeon         urban 350i  2012           Scooter               313.0   
361   ajp     pr4 125 enduro  2010  Enduro / offroad               124.0   
408   ajp  pr7 adventure 650  2018  Enduro / offroad               659.7   

     Power (hp)  Torque (Nm)  Engine cylinder Engine stroke    Gearbox  ...  \
195         3.0          3.7  Single cylinder    two-stroke  Automatic  ...   
203        20.1         30.0  Single cylinder   four-stroke  Automatic  ...   
226        22.8         30.0  Single cylinder   four-stroke  Automatic  ...   
361        12.5          8.5  Single cylinder   four-stroke    5-speed  ...   
408        48.0         58.0  Single cylinder   four-stroke    6-speed  ...   

     Dry weight (kg) Wheelbase (mm)  Seat height (mm)  \
195            129.0         1050.0             800.0   
203            236.0         1230.0             850.0   
226            177.0         1545.0             815.0   
361            105.0         1410.0             910.0   
408            155.0         1540.0             920.0   

                       Front brakes                         Rear brakes  \
195    Expanding brake (drum brake)                         Single disc   
203                     Double disc        Expanding brake (drum brake)   
226          Single disc. Hydraulic              Single disc. Hydraulic   
361  Single disc. 2 piston calliper      Single disc. 4 piston calliper   
408  Single disc. 2-piston calipers  Single disc. Single-piston caliper   

     Front tire   Rear tire                    Front suspension  \
195     19/7-8     18/10-8   Dual hydraulic shock, Single A-arm   
203    23/7-12    23/10-12                         Double A-Arm   
226  120/70-16   140/70-15                      Telescopic fork   
361   90/90-21   120/90-18               Paioli Hydraulic fork,   
408   90/90-21   140/80-18      ZF Sachs Ø48mm fully adjustable   

                                       Rear suspension         Color options  
195             Single hydraulic shock, Unit swing arm          White, black  
203                                          Swing Arm            Red, black  
226                         Dual-damper unit swing arm  White, black, silver  
361                Sachs mono shock progressive action                 Black  
408  ZF Sachs progressive system with reservatory f...       White/red/black  

[5 rows x 27 columns]

In [30]:
# Put our new 'Rating_' column into our dataframe
df = pd.concat([df,data], axis =1)

In [31]:
# Make sure it worked
df.head()

Brand Model  Year Category  Displacement (ccm)  Power (hp)  Torque (Nm)  \
0   NaN   NaN   NaN      NaN                 NaN         NaN          NaN   
1   NaN   NaN   NaN      NaN                 NaN         NaN          NaN   
2   NaN   NaN   NaN      NaN                 NaN         NaN          NaN   
3   NaN   NaN   NaN      NaN                 NaN         NaN          NaN   
4   NaN   NaN   NaN      NaN                 NaN         NaN          NaN   

  Engine cylinder Engine stroke Gearbox  ...  Wheelbase (mm) Seat height (mm)  \
0             NaN           NaN     NaN  ...             NaN              NaN   
1             NaN           NaN     NaN  ...             NaN              NaN   
2             NaN           NaN     NaN  ...             NaN              NaN   
3             NaN           NaN     NaN  ...             NaN              NaN   
4             NaN           NaN     NaN  ...             NaN              NaN   

   Front brakes Rear brakes Front tire Rear tire Front suspension  \
0           NaN         NaN        NaN       NaN              NaN   
1           NaN         NaN        NaN       NaN              NaN   
2           NaN         NaN        NaN       NaN              NaN   
3           NaN         NaN        NaN       NaN              NaN   
4           NaN         NaN        NaN       NaN              NaN   

   Rear suspension  Color options  Rating_  
0              NaN            NaN      0.0  
1              NaN            NaN      1.0  
2              NaN            NaN      1.0  
3              NaN            NaN      0.0  
4              NaN            NaN      1.0  

[5 rows x 28 columns]

In [32]:
# Preview data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3402 entries, 0 to 38298
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                1747 non-null   object 
 1   Model                1747 non-null   object 
 2   Year                 1747 non-null   float64
 3   Category             1747 non-null   object 
 4   Displacement (ccm)   1747 non-null   float64
 5   Power (hp)           1747 non-null   float64
 6   Torque (Nm)          1747 non-null   float64
 7   Engine cylinder      1747 non-null   object 
 8   Engine stroke        1747 non-null   object 
 9   Gearbox              1747 non-null   object 
 10  Bore (mm)            1747 non-null   float64
 11  Stroke (mm)          1747 non-null   object 
 12  Fuel capacity (lts)  1747 non-null   float64
 13  Fuel system          1747 non-null   object 
 14  Fuel control         1747 non-null   object 
 15  Cooling system       1747 non-null   

# Preprocessing data for train_test_split

## There's no replacement for displacement

In [ ]:
# 'Displacement (ccm)' would be how many cubic centimeters of engine displacement/size a bike's 
# engine has
print(df['Displacement (ccm)'].value_counts())
print()
print('There are {} null values in "Displacement (ccm)" '.format(df['Displacement (ccm)'].isnull().sum()))

<div class="alert alert-block alert-info">
Even though 8857 null values in 'Displacement (ccm)' would seem like a lot, engine displacement is an incredibly subjective parameter for a motorcycle. 
</div>

In [ ]:
# Drop null values for displacement
df = df.dropna(subset =['Displacement (ccm)'])

In [ ]:
# Preview dataset and identify what other work needs to be done
df.head()

<div class="alert alert-block alert-info">
Now that we have preprocessed one column in our dataset, let's take a look at what other columns we should consider processing. 
</div>

In [ ]:
# Number of null values in df['Displacement (ccm)']
df['Displacement (ccm)'].isnull().sum()

In [ ]:
# Preview data
df.info()

<div class="alert alert-block alert-warning">
There are a few considerations that we should make; there are 21684 total entries and we can assume that any column with fewer than that are null values. There are however a few things to keep in mind when rationalizing whether or not we should drop these values and ultimately, whether or not we should keep them at all. 
<br><br>    
Looking specifically at 'Torque (Nm)', this would imply that the motorcycle in question has at one point or another been put on a dyno to read out torque specs. While this is typically considered the "fun" value in a bike, let's first make a model with all null values in 'Torque (Nm' expunged. Once we have that model created and fitted, let's also look at a model that does not take into account torque figures. 
</div>



A dynamometer test is typically used to tell you the torque capabilities in your engine. Developed first in 1798, the technology has come a _long_ way since. <br><br>

<img src="pictures/DIY_dyno_complete1.jpg.crdownload" width = '500'/>
<center>Photo of DIY Dyno build from skrunkwerks.com</center><br><br>
Modern dynamometer testing is done by pushing air into the air filter (or turbo if you're on a drag monster) in order to simulate flow at speed and to push cool air into the radiator. By strapping the bike to the dyno, we are then able to get power readouts via sensors in the flywheel where the rear tire can put power to the device. <br><br>
I suspect that the bikes that do not have a torque reading either did not get tested due to extenuating circumstances (smaller displacement bikes, cruisers, and dirt bikes are usually not tested) or, the modern dynamometer was not available for some years of model submission. 

# Model with torque numbers intact

As suggested, we will first look at bikes that _have_ torque (nm) information. 

In [ ]:
# Remove null values from df['Torque (Nm)']
df = df.dropna(subset =['Torque (Nm)'])

In [ ]:
# Preview our data to see what other columns need pruning
df.info()

<div class="alert alert-block alert-warning">
There are a few considerations that we should in regards to null values and whether or not we believe these should be included in the first place. Since we have not made these values into numbers via encoding and therefore cannot yet see correlation to our target, responsibly we should only address lines with FULL information. <br><br> 
    
In other words, we should rid our dataset of null values prior to cutting out any column information. 
    
<br>
This isn't always the case, but for our purposes:
</div>

# If the value is null, it has to go

In [ ]:
# Drop null values data that we intend on using for modeling
df = df.dropna(subset =['Model', 'Power (hp)', 'Engine cylinder', 'Engine stroke', 'Gearbox', 'Bore (mm)', 'Stroke (mm)',
                       'Fuel capacity (lts)', 'Fuel system', 'Fuel control', 'Cooling system', 'Transmission type', 
                       'Dry weight (kg)', 'Wheelbase (mm)', 'Seat height (mm)', 'Front brakes', 'Rear brakes', 'Front tire',
                       'Rear tire', 'Front suspension', 'Rear suspension', 'Rating_'])

Because 'Color options' would seem to be a pretty objective thing and typically motorcycle manufacturers have up to 4-5 years to introducing new colorways per generation model of bike, this will ultimately be unnecessary for our model. 

In [ ]:
# Remove df['Color options']
df = df.drop(columns = 'Color options', axis = 1)

In [ ]:
# Preview our new dataset
df.head()

In [ ]:
# Information on our new dataset
df.info()

To round out our preprocessing step, we should take out any unnecessary information, like 'Brand' and 'Model'.  

In [ ]:
# Drop df['Brand']
df = df.drop(columns = 'Brand', axis =1)

In [ ]:
# Drop df['Model']
df = df.drop(columns = 'Model', axis =1)

<div class="alert alert-block alert-info">
Now that we have pruned all null information from our data set let's address remaining categorical data by dummying them. 

# Dummy categorical columns

Let's split our data into a categorical variable and dummy them into binary columns.

In [ ]:
# Categorical/object columns can be found from df.info(), above
categoricals = ['Category','Engine cylinder', 'Engine stroke', 'Gearbox', 'Stroke (mm)', 'Fuel system', 'Fuel control', 
                'Cooling system', 'Transmission type', 'Front brakes', 'Rear brakes', 'Front tire', 'Rear tire', 
                'Front suspension', 'Rear suspension']

# Our dataframe needs to be encoded. OHE can also be utilized, for this time we will use 
# pd.get_dummies()
df = pd.get_dummies(df, columns=categoricals)
# Preview our dummied columns
df.head()

<div class="alert alert-block alert-success">
It looks like our columns were successfully dummied. Now we will need to split our data for modeling. 
</div>

## train_test_split


In [ ]:
# target is 'y'
target = df['Rating_']
# X is the new name that we will use for our data. For testing purposes we may re-use 
#'df' if it produces better results.
X = df.drop(columns = 'Rating_', axis =1)

In [ ]:
#import train_test_split
from sklearn.model_selection import train_test_split

# Create variables for modeling
X_train, X_test, y_train, y_test = train_test_split (X, target, random_state = 42)

#  Baseline Model 1.0: 
First we will try a single DecisionTree

In [ ]:
#import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier 
dtc = DecisionTreeClassifier(random_state=42)  
# fit classifier onto training data
dtc.fit(X_train, y_train) 

In [ ]:
print('DecisionTree Model training score: {}'.format(dtc.score(X_train, y_train)))

In [ ]:
print('DecisionTree Model test score: {}'.format(dtc.score(X_test, y_test)))

<div class="alert alert-block alert-danger">
Our first model is built on a decision tree and the performance is not great. This is done with no regard to hyperparameters, so let's call this our baseline model. If that's the case, let's see how we can improve it. 
</div>

# KNN: K-Nearest-Neighbors
KNN or K-Nearest-Neighbors is a distance based estimator; we will need to scale our data prior to modeling it in order to get best results. 


In [ ]:
# This next step can be done without instantiating a pipeline but the tool makes the process simpler
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline


knn_pipeline = Pipeline([('ss', StandardScaler()),
                        ('knn', KNeighborsClassifier())])

In [ ]:
# Fit our knn_pipeline onto our training data, testing data and then see scores
knn_pipeline.fit(X_train,y_train)

print('KNN Training performance:{}'.format(knn_pipeline.score(X_train,y_train)))
print('KNN Test performance:{}'.format(knn_pipeline.score(X_test,y_test)))

Performance on our KNN model is not bad, but let's see if there is something that can give us a better test performance. 

# Bagged Decision Tree Model: 

Bagging is a form of ensemble model training. Bagging, the term is short for bootstrap aggregating.

In [ ]:
# import BaggingClassifier
from sklearn.ensemble import BaggingClassifier
b_tree =  BaggingClassifier(DecisionTreeClassifier(criterion='entropy', max_depth=5), 
                                 n_estimators=20, random_state = 42)
# fit model on training data
b_tree.fit(X_train, y_train)

In [ ]:
# Training data score
print('Bagged model training score: {}'.format(b_tree.score(X_train, y_train)))


In [ ]:
# Testing data score
print('Bagged model test score: {}'.format(b_tree.score(X_test, y_test)))


<div class="alert alert-block alert-success">
Our model has made improvements and does not appear to be overfitting. We can move forward with these parameters but first, let's see if GridSearchCV can build a better model. 
</div>

# DecisionTree Model 2.0: 

We will try using cross-validation to see if we can up the score of our base Decision Tree Model 1.0. Cross-validation assists with overfitting, and even though it did not appear that our model overfit the data, it is still worth a try.  

In [ ]:
# import cross_val_score from sklearn
from sklearn.model_selection import cross_val_score
# We will re-use 'dtc', our DecisionTreeClassifier() from earlier

dtc_cross_score = cross_val_score(dtc, X_train, y_train, cv = 3)
mean_dtc_cross_score = np.mean(dtc_cross_score)

print('Mean Cross-Validation Score:{}'.format(mean_dtc_cross_score))

<div class="alert alert-block alert-danger">
The performance on this model is worse than the first base DecisionTree that we put together. We will scrap this version for now and move forward with our Bagged Decision Tree. 
</div>

# Hyperparameter tuning on our First Decision Tree:

We will tune our DecisionTree Model using GridsearchCV. <br><br>
GridsearchCV is a very powerful tool that can tell us what tuning parameters are the best for this dataframe. 

In [ ]:
# Bring in hyperparameter keys for DecisionTreeClassifier
DecisionTreeClassifier().get_params().keys()

In [ ]:
# We should create a new param_grid for this section so that we can more easily 
# import GridSearchCV
from sklearn.model_selection import GridSearchCV


dtc_grid_search = GridSearchCV(estimator = (DecisionTreeClassifier(n_estimators=20, random_state = 42)), 
                               param_grid = {
                                   'criterion': ['gini', 'entropy','log_loss'],
                                   'max_depth': [10,20,30],
                                   'random_state': [42]
                               }, 
                                  cv = 3
                                  )

#Fit GridSearchCV function to our data
dtc_grid_search.fit(X_train, y_train)

In [ ]:
# Our mean training score
dtc_training_score = dtc_grid_search.score(X_train, y_train)

# Mean Test score
dtc_testing_score = dtc_grid_search.score(X_test, y_test)

print(f"Mean Training Score: {dtc_training_score :.2%}")
print(f"Mean Test Score: {dtc_testing_score :.2%}")
print("Best Parameter Combination Found During Grid Search:")
dtc_grid_search.best_params_

# Hyperparameter tuning on our Bagged Decision Tree:
Since we tuned the hyperparameters on our worst performing model, let's see if we can improve the score on our best performing model. 

In [ ]:
# Since we are hyperparameter tuning a BaggingClassifier for the first time, let's see what exactly
# we are allowed to tune. Further reading : 
# https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html
from sklearn.ensemble import RandomForestClassifier
RandomForestClassifier().get_params().keys()

In [ ]:
classifier = BaggingClassifier(DecisionTreeClassifier(criterion='entropy', max_depth=5), 
                                 n_estimators=20, random_state = 42)

In [ ]:
# For GridSearchCV, we can set possible parameters in this cell and fit GridsearchCV to cross-validate
# all possible parameters for best outcome. 

# param_grid = {
# #     'base_estimator': [None], # This will default to a DecisionTreeClassifier
#     'bootstrap': [True], # Default option, true bootstrapping is done by replacement
#     'bootstrap_features': [True], #Default option
#     'max_features': [1,2,3], # This is the number of base_estimators to train from X
#     'max_samples': [1,2,3,4], #Number of sample to draw from X to train each base_estimator
#     'n_estimators': [20,50,100], #Number of base_estimators in ensemble
# #     'n_jobs': [-1], # -1 uses all processors to fit and predict on model
#     'oob_score': [True,False], # Whether or not to use 'out-of-bag' samples or not
#     'random_state': [42], # The answer to the universe
#     'verbose': [0], # Unsure what verbosity is, looking to see what happens
#     'warm_start': [True,False] # When 'True', reuses previous solution and adds estimators to ensemble
    
# }

In [ ]:
# Fun Fact: This will take around 3-4 minutes to run, 'Alright' by Kenrick Lamar is a good choice

# We can re-use our Bagged DecisionTree, b_tree to improve performance on our Bagged DecisionTree
# import GridSearchCV


# Instantiate GridSearchCV
b_tree_grid_search = GridSearchCV(bag_classifier, #bag_classifier is our bagged DecisionTree
                                  param_grid = {
                                   'criterion': ['gini', 'entropy','log_loss'],
                                   'max_depth': [10,20,30],
                                   'random_state': [42]
                               }, 
                                  cv = 3 
                                  )

#Fit GridSearchCV function to our data
b_tree_grid_search.fit(X_train, y_train)

In [ ]:
# Our mean training score
b_tree_training_score = b_tree_grid_search.score(X_train, y_train)

# Mean Test score
b_tree_testing_score = b_tree_grid_search.score(X_test, y_test)

print(f"Mean Training Score: {b_tree_training_score :.2%}")
print(f"Mean Test Score: {b_tree_testing_score :.2%}")
print("Best Parameter Combination Found During Grid Search:")
b_tree_grid_search.best_params_